File: Raven.Abstractions/Json/ConflictsResolver.cs
using System;
using System.Collections.Generic;
using System.IO;
using Raven.Imports.Newtonsoft.Json;
using Raven.Imports.Newtonsoft.Json.Linq;
using Raven.Json.Linq;

namespace Raven.Abstractions.Json
{
	public class ConflictsResolver
	{
		private readonly RavenJObject[] docs;
		readonly RavenJTokenEqualityComparer ravenJTokenEqualityComparer = new RavenJTokenEqualityComparer();

		public ConflictsResolver(params RavenJObject[] docs)
		{
			this.docs = docs;
		}

		public string Resolve(int indent = 1)
		{
			var result = new Dictionary<string, object>();
			for (int index = 0; index < docs.Length; index++)
			{
				var doc = docs[index];
				foreach (var prop in doc)
				{
					if (result.ContainsKey(prop.Key)) // already dealt with
						continue;

					switch (prop.Value.Type)
					{
						case JTokenType.Object:
							if(TryHandleObjectValue(index, result, prop) == false)
								goto default;
							break;
						case JTokenType.Array:
							if(TryHandleArrayValue(index, result, prop) == false)
								goto default;
							break;
						default:
							HandleSimpleValues(result, prop, index);
							break;
					}
				}
			}
			return GenerateOutput(result, indent);
		}

		private bool TryHandleArrayValue(int index, Dictionary<string, object> result, KeyValuePair<string, RavenJToken> prop)
		{
			var arrays = new List<RavenJArray>
			{
				(RavenJArray)prop.Value
			};
			for (int i = 0; i < docs.Length; i++)
			{
				if (i == index)
					continue;

				RavenJToken token;
				if (docs[i].TryGetValue(prop.Key, out token) && token.Type != JTokenType.Array)
					return false;
				if (token == null)
					continue;
				arrays.Add((RavenJArray)token);
			}

			var mergedArray = new RavenJArray();
			while (arrays.Count > 0)
			{
				var set = new HashSet<RavenJToken>(ravenJTokenEqualityComparer);
				for (int i = 0; i < arrays.Count; i++)
				{
					set.Add(arrays[i][0]);
					arrays[i].RemoveAt(0);
					if(arrays[i].Length == 0)
						arrays.RemoveAt(i);
				}
				foreach (var ravenJToken in set)
				{
					mergedArray.Add(ravenJToken);
				}
			}

			if (ravenJTokenEqualityComparer.Equals(mergedArray, prop.Value))
			{
				result.Add(prop.Key, mergedArray); 
				return true;
			}

			result.Add(prop.Key, new ArrayWithWarning(mergedArray));
			return true;
		}

		private bool TryHandleObjectValue(int index, Dictionary<string, object> result, KeyValuePair<string, RavenJToken> prop)
		{
			var others = new List<RavenJObject>
			{
				(RavenJObject)prop.Value
			};
			for (int i = 0; i < docs.Length; i++)
			{
				if(i == index)
					continue;

				RavenJToken token;
				if (docs[i].TryGetValue(prop.Key, out token) && token.Type != JTokenType.Object)
					return false;
				if(token == null)
					continue;
				others.Add((RavenJObject)token);
			}
			result.Add(prop.Key, new ConflictsResolver(others.ToArray()));
			return true;
		}

		private void HandleSimpleValues(Dictionary<string, object> result,
			KeyValuePair<string, RavenJToken> prop, 
			int index)
		{
			var conflicted = new Conflicted
			{
				Values = { prop.Value }
			};
			for (int i = 0; i < docs.Length; i++)
			{
				if(i == index)
					continue;
				var other = docs[i];
			
				RavenJToken otherVal;
				if (other.TryGetValue(prop.Key, out otherVal) == false)
					continue;
				if (ravenJTokenEqualityComparer.Equals(prop.Value, otherVal) == false)
					conflicted.Values.Add(otherVal);
			}
			if (conflicted.Values.Count == 1)
				result.Add(prop.Key, prop.Value);
			else
				result.Add(prop.Key, conflicted);
		}

		private static string GenerateOutput(Dictionary<string, object> result, int indent)
		{
			var stringWriter = new StringWriter();
			var writer = new JsonTextWriter(stringWriter)
			{
				Formatting = Formatting.Indented
			};
			writer.WriteStartObject();
			foreach (var o in result)
			{
				writer.WritePropertyName(o.Key);
				var ravenJToken = o.Value as RavenJToken;
				if (ravenJToken != null)
				{
					ravenJToken.WriteTo(writer);
					continue;
				}
				var conflicted = o.Value as Conflicted;
				if (conflicted != null)
				{
					writer.WriteComment(">>>> conflict start");
					writer.WriteStartArray();
					foreach (var token in conflicted.Values)
					{
						token.WriteTo(writer);
					}
					writer.WriteEndArray();
					writer.WriteComment("<<<< conflict end");
					continue;
				}
				var arrayWithWarning = o.Value as ArrayWithWarning;
				if(arrayWithWarning != null)
				{
					writer.WriteComment(">>>> auto merged array start");
					arrayWithWarning.MergedArray.WriteTo(writer);
					writer.WriteComment("<<<< auto merged array end");
					continue;
				}
				var resolver = o.Value as ConflictsResolver;
				if(resolver != null)
				{
					using(var stringReader = new StringReader(resolver.Resolve(indent + 1)))
					{
						bool first = true;
						string line ;
						while((line = stringReader.ReadLine()) != null)
						{
							if(first == false)
							{
								writer.WriteRaw(Environment.NewLine);
								for (var i = 0; i < indent; i++)
								{
									writer.WriteRaw(new string(writer.IndentChar, writer.Indentation));
								}
							}
							first = false;
							writer.WriteRaw(line);
						}
					}
					continue;
				}
				throw new InvalidOperationException("Could not understand how to deal with: " + o.Value);
			}
			writer.WriteEndObject();
			return stringWriter.GetStringBuilder().ToString();
		}

		private class Conflicted
		{
			public readonly HashSet<RavenJToken> Values = new HashSet<RavenJToken>(new RavenJTokenEqualityComparer());
		}

		private class ArrayWithWarning
		{
			public readonly RavenJArray MergedArray;

			public ArrayWithWarning(RavenJArray mergedArray)
			{
				this.MergedArray = mergedArray;
			}
		}

	}

}

File: Raven.Client.Java/src/main/java/net/ravendb/abstractions/json/ConflictsResolver.java
package net.ravendb.abstractions.json;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.StringReader;
import java.io.StringWriter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.TreeSet;

import org.codehaus.jackson.JsonFactory;
import org.codehaus.jackson.JsonGenerator;

import net.ravendb.abstractions.json.linq.JTokenType;
import net.ravendb.abstractions.json.linq.RavenJArray;
import net.ravendb.abstractions.json.linq.RavenJObject;
import net.ravendb.abstractions.json.linq.RavenJToken;
import net.ravendb.abstractions.json.linq.RavenJTokenComparator;


public class ConflictsResolver {

  private RavenJObject[] docs;
  private RavenJTokenComparator ravenJTokenComparator = new RavenJTokenComparator();

  public ConflictsResolver(RavenJObject... docs) {
    this.docs = docs;
  }

  public String resolve() {
    return resolve(1);
  }

  public String resolve(int indent) {
    Map<String, Object> result = new HashMap<>();
    for (int index = 0; index < docs.length; index++) {
      RavenJObject doc = docs[index];
      for (Entry<String, RavenJToken> prop : doc) {
        if (result.containsKey(prop.getKey())) { // already dealt with
          continue;
        }

        JTokenType type = prop.getValue().getType();
        boolean executeDefault = true;
        if (type == JTokenType.OBJECT && tryHandleObjectValue(index, result, prop)) {
          executeDefault = false;
        }
        if (type == JTokenType.ARRAY && tryHandleArrayValue(index, result, prop)) {
          executeDefault = false;
        }
        if (executeDefault) {
          handleSimpleValues(result, prop, index);
        }
      }
    }
    return generateOutput(result, indent);
  }

  private boolean tryHandleArrayValue(int index, Map<String, Object> result, Map.Entry<String, RavenJToken> prop) {
    List<RavenJArray> arrays = new ArrayList<>();
    arrays.add((RavenJArray) prop.getValue());

    for (int i = 0; i < docs.length; i++) {
      if (i == index) {
        continue;
      }

      RavenJToken token = null;
      if (docs[i].containsKey(prop.getKey())) {
        token = docs[i].get(prop.getKey());
        if (token.getType() != JTokenType.ARRAY) {
          return false;
        }
        arrays.add((RavenJArray) token);
      }
    }

    RavenJArray mergedArray = new RavenJArray();
    while (arrays.size() > 0) {
      Set<RavenJToken> set = new TreeSet<>(ravenJTokenComparator);
      for (int i = 0; i < arrays.size(); i++) {
        if (arrays.get(i).size() == 0) {
          arrays.remove(i);
          i--;
          continue;
        }
        set.add(arrays.get(i).get(0));
        arrays.get(i).removeAt(0);
      }

      for (RavenJToken ravenJToken : set) {
        mergedArray.add(ravenJToken);
      }
    }

    if (ravenJTokenComparator.compare(mergedArray, prop.getValue()) == 0) {
      result.put(prop.getKey(), mergedArray);
      return true;
    }

    result.put(prop.getKey(), new ArrayWithWarning(mergedArray));
    return true;

  }

  private boolean tryHandleObjectValue(int index, Map<String, Object> result, Map.Entry<String, RavenJToken> prop) {
    List<RavenJObject> others = new ArrayList<>();
    others.add((RavenJObject) prop.getValue());

    for (int i = 0; i < docs.length; i++) {
      if (i == index) {
        continue;
      }

      RavenJToken token = null;
      if (docs[i].containsKey(prop.getKey())) {
        token = docs[i].get(prop.getKey());
        if (token.getType() != JTokenType.OBJECT) {
          return false;
        }
        others.add((RavenJObject)token);
      }
    }
    result.put(prop.getKey(), new ConflictsResolver(others.toArray(new RavenJObject[0])));
    return true;
  }


  private void handleSimpleValues(Map<String, Object> result, Map.Entry<String, RavenJToken> prop, int index) {
    Conflicted conflicted = new Conflicted();
    conflicted.getValues().add(prop.getValue());

    for (int i = 0; i < docs.length; i++) {
      if (i == index) {
        continue;
      }
      RavenJObject other = docs[i];
      RavenJToken otherVal = null;
      if (!other.containsKey(prop.getKey())) {
        continue;
      }
      otherVal = other.get(prop.getKey());

      if (ravenJTokenComparator.compare(prop.getValue(), otherVal) != 0) {
        conflicted.getValues().add(otherVal);
      }
    }
    if (conflicted.getValues().size() == 1) {
      result.put(prop.getKey(), prop.getValue());
    } else {
      result.put(prop.getKey(), conflicted);
    }
  }

  private static String generateOutput(Map<String, Object> result, int indent) {
    try {
      JsonFactory factory = new JsonFactory();
      StringWriter stringWriter = new StringWriter();
      JsonGenerator writer = factory.createJsonGenerator(stringWriter);

      writer.writeStartObject();
      for (Map.Entry<String, Object> o : result.entrySet()) {
        writer.writeFieldName(o.getKey());
        if (o.getValue() instanceof RavenJToken) {
          RavenJToken ravenJToken = (RavenJToken) o.getValue();
          ravenJToken.writeTo(writer);
          continue;
        }
        if (o.getValue() instanceof Conflicted) {
          Conflicted conflicted = (Conflicted) o.getValue();
          writer.writeRaw("/* >>>> conflict start */");
          writer.writeStartArray();
          for (RavenJToken token : conflicted.getValues()) {
            token.writeTo(writer);
          }
          writer.writeEndArray();
          writer.writeRaw("/* <<<< conflict end */");
          continue;
        }

        if (o.getValue() instanceof ArrayWithWarning) {
          ArrayWithWarning arrayWithWarning = (ArrayWithWarning) o.getValue();
          writer.writeRaw("/* >>>> auto merged array start */");
          arrayWithWarning.getMergedArray().writeTo(writer);
          writer.writeRaw("/* <<<< auto merged array end */");
          continue;
        }

        if (o.getValue() instanceof ConflictsResolver) {
          ConflictsResolver resolver = (ConflictsResolver) o.getValue();

          BufferedReader stringReader = new BufferedReader(new StringReader(resolver.resolve(indent + 1)));
          boolean first = true;
          String line = null;
          while ((line = stringReader.readLine()) != null) {
            if (!first) {
              writer.writeRaw(System.lineSeparator());
              for (int i = 0; i < indent; i++) {
                writer.writeRaw(" ");
              }
            }
            if (first) {
              writer.writeRawValue(line);
            } else {
              writer.writeRaw(line);
            }

            first = false;
          }

          continue;
        }

        throw new IllegalStateException("Could not understand how to deal with: " + o.getValue());

      }
      writer.writeEndObject();
      writer.close();
      return stringWriter.getBuffer().toString();
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }


  public static class Conflicted {
    private final Set<RavenJToken> values = new TreeSet<>(new RavenJTokenComparator());

    public Set<RavenJToken> getValues() {
      return values;
    }


  }

  public static class ArrayWithWarning {
    private final RavenJArray mergedArray;

    public RavenJArray getMergedArray() {
      return mergedArray;
    }

    public ArrayWithWarning(RavenJArray mergedArray) {
      super();
      this.mergedArray = mergedArray;
    }
  }
}
